Smart coloring

victorviola opened the pull request:
Added a new feature to paint the diagram. This feature has the following behavior:

- A new button to force painting the whole project.
- Smart Coloring

Smart coloring is a functionality to paint the model by the time the user is working on it. It was possible to achieve this behavior by the implementation of listeners available in the OpenAPI. Now the plugin infers the color of the elements based on the generalizations of the classes (directly and indirectly relations).

claudenirmf commented:
I have faced a NullPointerException on the ModelListenner when trying to mess with the model. My guess is that it was triggered when trying to run on a generalization that I was deleting and that had null as return of `getFrom()` and `getTo()`. Follows the exception: 

```
java.lang.NullPointerException
        at it.unibz.inf.ontouml.vp.listeners.ModelListener.propertyChange(ModelListener.java:39)
        at v.beo.bvd.propertyChange(:32)
        at java.beans.PropertyChangeSupport.fire(PropertyChangeSupport.java:335)
        at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:327)
        at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:263)
        at v.chq.bte.r_firePropertyChange(:3133)
        at v.chq.bol.q(:1502)
        at v.chq.bol.setModelPropertyAsString(:656)
        at v.chq.in.setModelPropertyAsString(:555)
        at v.chq.bte.updatePmLastModified(:4580)
        at v.chq.bte.r_save(:4527)
        at v.chq.byo.r_save(:611)
        at v.chq.bot.r_save(:420)
        at v.chq.bol.r_save(:85)
        at v.chq.in.r_save(:114)
        at v.chq.bte.saveModelWithCompositeChild(:4452)
        at v.chq.bte.saveModel(:4441)
        at v.ckz.al.a(:328)
        at v.ckz.al.b(:202)
        at v.ckz.ak.a(:521)
        at v.ckz.ak.a(:509)
        at v.cgc.du.b(:5983)
        at v.cgc.du.a(:5948)
        at v.avd.aby.a(:11329)
        at v.avd.w.run(:54)
```

claudenirmf commented:
I don't like that the button for smart paint is enabled while the feature is disabled in the Settings menu. If you agree, could you please show a dialog on the screen (or some other warning) when the user has the feature disabled and clicks on Smart Paint?

Otherwise, the user will have no feedback on why the button is doing nothing. A part of me would like to have a confirmation on the screen even when the feature is working, just to provide the user some feedback, but I'm afraid of becoming annoying with that.

claudenirmf commented on src/main/java/it/unibz/inf/ontouml/vp/utils/SmartColoring.java:
```diff
@@ -0,0 +1,276 @@
+
+	/**
+	 * 
+	 * Runs twice over the diagram and paint all the elements.
```
It seems to run only once over the whole model.

victorviola commented:
Hi @claudenirmf 

1 - I have added a fix to handle the null pointer exception when deleting a generalization
2 - I have added a warning message in the bottom of the editor when trying to smart paint while the auto coloring is disabled (If you guys still prefer dialogs I can change that.)
3 - That inner for runs twice over the model

claudenirmf commented on src/main/java/it/unibz/inf/ontouml/vp/controllers/PaintModelAction.java:
```diff
@@ -1,14 +1,25 @@
 	public void performAction(VPAction arg0) {
+		
+		if (!Configurations.getInstance().getProjectConfigurations().isAutomaticColoringEnabled()) {
+			_viewManager.showMessage("Smart Paint is NOT enabled! Please activate it on the plugin settings.");		
```
If the team is happy with the message in the log pane, I am fine with it.
If we decide, instead, to use a warning dialog, we can replace this line with the following code:
```java
_viewManager.showConfirmDialog(null, 
      "Smart Paint is NOT enabled! Please activate it on the plugin settings.", 
      "Smart Paint", 
      JOptionPane.DEFAULT_OPTION, 
      JOptionPane.INFORMATION_MESSAGE);
```

We could even add an icon to replace Java's coffee mug icon.

Beyond this, I believe we can accept the PR.
